x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,287 @@
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+//_________________________________________________________________________
+// Accumulated updates to the active/bonded validator set for tendermint
+
+// get the most recently updated validators
+//
+// CONTRACT: Only validators with non-zero power or zero-power that were bonded
+// at the previous block height or were removed from the validator set entirely
+// are returned to Tendermint.
+func (k Keeper) GetTendermintUpdates(ctx sdk.Context) (updates []abci.ValidatorUpdate) {
+
+	store := ctx.KVStore(k.storeKey)
+	maxValidators := k.GetParams(ctx).MaxValidators
+
+	// copy last validator set
+	last := make(map[[sdk.AddrLen]byte][]byte)
+	iterator := sdk.KVStorePrefixIterator(store, BondedValidatorsIndexKey)
+	for ; iterator.Valid(); iterator.Next() {
+		var operator [sdk.AddrLen]byte
+		copy(operator[:], iterator.Key()[1:])
+		powerBytes := iterator.Value()
+		last[operator] = make([]byte, len(powerBytes))
+		copy(last[operator][:], powerBytes[:])
+	}
+
+	iterator = sdk.KVStoreReversePrefixIterator(store, ValidatorsByPowerIndexKey)
+	count := 0
+	for ; iterator.Valid() && count < int(maxValidators); iterator.Next() {
+
+		// fetch the validator
+		operator := sdk.ValAddress(iterator.Value())
+		validator := k.mustGetValidator(ctx, operator)
+
+		// jailed validators are ranked last, so if we get to a jailed validator
+		// we have no more bonded validators
+		// TODO we can remove this if we remove jailed validators from the power store
+		// likewise for zero-power validators, which we never bond
+		if validator.Jailed || validator.Tokens.Equal(sdk.ZeroDec()) {
+			break
+		}
+
+		// apply the appropriate state change if necessary
+		switch validator.Status {
+		case sdk.Unbonded:
+			validator = k.unbondedToBonded(ctx, validator)
+		case sdk.Unbonding:
+			validator = k.unbondingToBonded(ctx, validator)
+		case sdk.Bonded:
+			// no state change
+		}
+
+		// validator still in the validator set
+		var opbytes [sdk.AddrLen]byte
+		copy(opbytes[:], operator[:])
+
+		// fetch the old power bytes
+		powerBytes, ok := last[opbytes]
+
+		// calculate the new power bytes
+		newPowerBytes := validator.ABCIValidatorPowerBytes(k.cdc)
+
+		// update the validator set if power has changed
+		if !ok || !bytes.Equal(powerBytes, newPowerBytes) {
+			updates = append(updates, validator.ABCIValidatorUpdate())
+		}
+
+		// validator still in the validator set
+		delete(last, opbytes)
+
+		// set the bonded validator index
+		store.Set(GetBondedValidatorIndexKey(operator), newPowerBytes)
+
+		// keep count
+		count++
+
+	}
+
+	// sort the map keys for determinism
+	noLongerBonded := make([][]byte, len(last))
+	index := 0
+	for oper := range last {
+		operator := make([]byte, sdk.AddrLen)
+		copy(operator[:], oper[:])
+		noLongerBonded[index] = operator
+		index++
+	}
+	sort.SliceStable(noLongerBonded, func(i, j int) bool {
```

Comment:
Here we're sorting by power, correct? Suggest we add a short comment about that.

Reply:
No, we're sorting by address. Order shouldn't matter here but safer to be deterministic. Added a comment.